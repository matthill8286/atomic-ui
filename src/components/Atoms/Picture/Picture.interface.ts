export type PictureObjectFit = 'none' | 'contain' | 'cover' | 'fill' | 'scale-down'

export interface PictureProps {
  width?: string
  height?: string
  src?: string | null
  srcLg?: string | null
  srcMd?: string | null
  srcSm?: string | null
  alt?: string | null
  objectFit?: PictureObjectFit
  rounded?: boolean
  onClick?: (e: React.MouseEvent<HTMLElement>) => void
  onMouseEnter?: () => void
  onMouseLeave?: () => void
  disabled?: boolean
  loading?: NativeLazyLoadType
}

export enum NativeLazyLoadOptions {
  Eager = 'eager',
  Lazy = 'lazy',
}

export type NativeLazyLoadType = NativeLazyLoadOptions | undefined

export interface StyledPictureProps {
  width?: string
  height?: string
  imageMaxHeight?: string
  imageMaxWidth?: string
  objectFit?: PictureObjectFit
  rounded?: boolean
  disabled?: boolean
}
