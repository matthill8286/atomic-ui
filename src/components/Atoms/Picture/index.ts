export * from './Picture'
export * from './Picture.interface'
