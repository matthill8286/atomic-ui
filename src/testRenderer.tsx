import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
