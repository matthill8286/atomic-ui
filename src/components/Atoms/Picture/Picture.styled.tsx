import { media } from '@/styles/media'
import { css, styled } from '@/styles/styled'
import { StyledPictureProps } from './Picture.interface'

const disabledPicture = (isDisabled: boolean) => {
  if (isDisabled) {
    return css`
      -ms-filter: grayscale(100%);
      filter: grayscale(100%);
      opacity: 0.4;
      pointer-events: none;
    `
  }

  return ''
}

export const StyledPicture = styled.div<StyledPictureProps>`
  position: relative;
  display: flex;
  width: 100%;
  max-width: 100%;

  ${({ disabled }) => disabledPicture(!!disabled)};

  ${media.ie11} {
    max-width: ${({ width }) => (width ? `${width}` : '100%')};
  }

  & img {
    ${({ rounded, theme }) => (rounded ? `border-radius: ${theme.dimension.borderRadius2};` : '')};
    ${({ imageMaxHeight }) => (imageMaxHeight ? `max-height: ${imageMaxHeight};` : '')};
    ${({ imageMaxWidth }) => (imageMaxWidth ? `max-width: ${imageMaxWidth};` : '')};
    ${({ height }) => (height ? `height: ${height};` : '')};
    ${({ objectFit }) => (objectFit ? `object-fit: ${objectFit};` : '')};

    ${media.ie11} {
      max-width: 100%;
      width: auto;
      height: auto;
    }
  }
`
