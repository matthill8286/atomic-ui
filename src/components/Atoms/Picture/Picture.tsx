import React from 'react'
import { breakpoints } from '@/styles'
import { NativeLazyLoadOptions, PictureObjectFit, PictureProps } from './Picture.interface'
import { StyledPicture } from './Picture.styled'
import { SkeletonBlockItem } from '@/components/Atoms/Skeleton'

export const Picture: React.FC<PictureProps> = ({
  width = '100%',
  height,
  src,
  srcLg,
  srcMd,
  srcSm,
  alt,
  children,
  objectFit,
  rounded,
  loading = NativeLazyLoadOptions.Eager,
  ...otherProps
}): React.ReactElement => {
  return (
    <StyledPicture
      width={width}
      height={height}
      objectFit={(objectFit as PictureObjectFit) || 'cover'}
      rounded={rounded}
      {...otherProps}>
      <picture style={{ width: '100%', height }}>
        {srcSm && <source media={`(max-width: ${breakpoints.sm - 1}px)`} srcSet={srcSm} />}
        {srcMd && <source media={`(max-width: ${breakpoints.md - 1}px)`} srcSet={srcMd} />}
        {srcLg && <source media={`(max-width: ${breakpoints.lg - 1}px)`} srcSet={srcLg} />}
        {src ? (
          <img src={src} alt={alt || 'image'} width={width} height={height} loading={loading} />
        ) : (
          <SkeletonBlockItem width={width} height={height} />
        )}
      </picture>
      {children}
    </StyledPicture>
  )
}
