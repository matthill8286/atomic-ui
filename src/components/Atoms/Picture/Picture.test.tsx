import React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { Picture } from './Picture'

const IMAGE_URL =
  'https://learn.saiyan.com/hubfs/magpie%20client%20images%20and%20files/Procter%20and%20Gamble/instead%20of%20coming%20soon.png'

describe('Picture', () => {
  it('renders correctly', () => {
    const tree = renderWithTheme(<Picture src={IMAGE_URL} alt="Lorem ipsum" />)
    expect(tree).toMatchSnapshot()
  })
})
