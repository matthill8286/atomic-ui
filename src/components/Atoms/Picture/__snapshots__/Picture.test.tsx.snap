// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Picture renders correctly 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  max-width: 100%;
}

.c0 img {
  object-fit: cover;
}

@media screen and (-ms-high-contrast:active),screen and (-ms-high-contrast:none) {
  .c0 {
    max-width: 100%;
  }
}

@media screen and (-ms-high-contrast:active),screen and (-ms-high-contrast:none) {
  .c0 img {
    max-width: 100%;
    width: auto;
    height: auto;
  }
}

<div
  className="c0"
  width="100%"
>
  <picture
    style={
      Object {
        "height": undefined,
        "width": "100%",
      }
    }
  >
    <img
      alt="Lorem ipsum"
      loading="eager"
      src="https://learn.filtered.com/hubfs/magpie%20client%20images%20and%20files/Procter%20and%20Gamble/instead%20of%20coming%20soon.png"
      width="100%"
    />
  </picture>
</div>
`;
